### PR TITLE
Support converting hf gemma weights

### PR DESCRIPTION
**Pull request by lsy323**
- Update `convert_checkpoint.py` to convert Gemma weights from HuggingFace to `safetensor` format, and PyTorch `state_dict`.
- Enable loading converted Gemma weights in both safetensor format and `state_dict`
- Update `README.md` for running Gemma

Tested with running `convert_checkpoint.py` and `run_interactive.py`

### CODE REVIEW

File: convert_checkpoints.py [hunk 1]
```diff
@@ -403,16 +404,70 @@ def merge_weights(
...
+      state_dict[new_key.replace("qkv_proj", "wv")] = v
+      continue
+    if "o_proj" in key:
+      new_key = new_key.replace("o_proj", "wo")
+    if new_key != key:
+      state_dict[new_key] = state_dict.pop(key)
+
+  ckpt_basename = os.path.basename(ckpt_file)
+  output_ckpt_dir.mkdir(parents=True, exist_ok=True)
+  torch.save(
```
Comment (qihqi) [inline convert_checkpoints.py:449]: let's still use _export_to_local function here to produce safetensors here

File: convert_checkpoints.py [hunk 2]
```diff
@@ -403,16 +404,70 @@ def merge_weights(
...
+  1. Map attention weights to new names.
+  2. Split qkv fusion.
+  """
+  ckpt_file = list(input_ckpt_dir.glob("*.ckpt"))
+  assert len(ckpt_file) == 1
+  ckpt_file = ckpt_file[0]
+  state_dict = torch.load(ckpt_file, map_location=torch.device("cpu"))[
+      "model_state_dict"
+  ]
+  model_config = json.loads((input_ckpt_dir / "config.json").read_text())
```
Comment (qihqi) [inline convert_checkpoints.py:420]: you can change `json.load` with `eval` to skip the need to change `'` to `"` in the instructions.

File: run_interactive.py [hunk 3]
```diff
@@ -72,11 +72,6 @@
 _MAX_CACHE_LENGTH = flags.DEFINE_integer(
     "max_cache_length", 1024, "kv_cache_quantize"
 )
-_MODEL_NAME = flags.DEFINE_string(
```
[the remark this answers is not in the document]
Comment (lsy323) [reply run_interactive.py:75]: It's a duplicated flag, probably it got in with other PRs

**Comment by qihqi** [comment]
Can you also add the instructions in your PR to the README.md? thanks!

File: convert_checkpoints.py [hunk 4]
```diff
@@ -403,16 +404,70 @@ def merge_weights(
...
+      continue
+    if "o_proj" in key:
+      new_key = new_key.replace("o_proj", "wo")
+    if new_key != key:
+      state_dict[new_key] = state_dict.pop(key)
+
+  ckpt_basename = os.path.basename(ckpt_file)
+  output_ckpt_dir.mkdir(parents=True, exist_ok=True)
+  torch.save(
+      {"model_state_dict": state_dict},
```
Comment (qihqi) [inline convert_checkpoints.py:450]: remove the "model_state_dict" key; just save the state_dict;
then engine dont need to remove it

File: convert_checkpoints.py (hunk 2 above)
Comment (lsy323) [reply convert_checkpoints.py:420]: If using `eval`, we need to change `false` to `False` in the `config.json`

File: convert_checkpoints.py [hunk 5]
```diff
@@ -403,16 +404,70 @@ def merge_weights(
   print(f"Export outputs takes {end - start} seconds")
 
 
+def convert_hf_gemma_weights(
```
Comment (FanhaiLu1) [inline convert_checkpoints.py:407]: To be consistent with llama model, can we also convert gemma weights as safetensors?

Comment (FanhaiLu1) [reply convert_checkpoints.py:407]: Code can also be re-used from convert and load if we use safe tensors format.

File: run_interactive.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (FanhaiLu1) [reply run_interactive.py:75]: Thanks for removing, I planed to removed in another PR before review this PR (right now, I only need to pull your request later).

The first _MODEL_NAME  reads from --model_name, and this one reads from model. They are not exactly same and confused.

**Review by FanhaiLu1 — COMMENTED** [review]
Can you share the real output result with the gemma weights?

File: convert_checkpoints.py (hunk 2 above)
Comment (qihqi) [reply convert_checkpoints.py:420]: or let's read the config from https://github.com/google/jetstream-pytorch/blob/main/jetstream_pt/third_party/gemma/config.py#L81 using --size argument?

File: convert_checkpoints.py (hunk 5 above)
Comment (lsy323) [reply convert_checkpoints.py:407]: For `google/gemma-7b-pytorch` it contains `torch.complex64` dtype in the weight, which is not supported by safetensors. We have to convert to pytorch state dict format for loading.

Comment (lsy323) [reply convert_checkpoints.py:407]: I'll reuse the existing function to save weights in either safetensor or state_dict format, and add assertion to disallow safetensor output format for gemma model for now.

**Comment by lsy323** [comment]
> Can you share the real output result with the gemma weights?

There is some issue with the response, it's generating English but it will repeat the sentence. @qihqi will update the attention module for Gemma, which should fix the accuracy issue.

Prompt
```
"I believe the meaning of life is"
```
The response is:
```
 to experience.

Here is my reasoning:

* **Life is a journey.** We are all on a journey through time. The experiences we gain through life's journey is to experience the world and all that we experience, both positive and negative experiences, both good and bad experiences, both positive and negative experiences, both joy and suffering. We all go through life to experience and learn from both good and bad and positive and negative and positive and negative experiences, both joy and suffering and pain and joy, both good and bad and good and bad and negative and positive and negative and positive and negative and good and bad and the good and bad and negative and positive and negative and positive and negative and the good and bad and negative and positive and negative experiences, both good and bad and the good and bad and negative and the good and bad. The is the good and bad and the good and bad and the the good and bad and the the are the experiences. Through the joy and negative. The the the journey is the good. The through the the.
```

File: convert_checkpoints.py (hunk 2 above)
Comment (lsy323) [reply convert_checkpoints.py:420]: I think it's better to submit a PR to huggingface to correct the error. It should be an valid JSON. Not all Gemma ckpt on HF have this issue.

File: convert_checkpoints.py (hunk 5 above)
Comment (FanhaiLu1) [reply convert_checkpoints.py:407]: Can you share which items are torch.complex64? If it's freqs_cis, we can ignore it and get it from pytorch model.

**Comment by FanhaiLu1** [comment]
> > Can you share the real output result with the gemma weights?
> 
> There is some issue with the response, it's generating English but it will repeat the sentence. @qihqi will update the attention module for Gemma in the following PR, which should fix the accuracy issue.
> 
> Prompt
> 
> ```
> "I believe the meaning of life is"
> ```
> 
> The response is:
> 
> ```
>  to experience.
> 
> Here is my reasoning:
> 
> * **Life is a journey.** We are all on a journey through time. The experiences we gain through life's journey is to experience the world and all that we experience, both positive and negative experiences, both good and bad experiences, both positive and negative experiences, both joy and suffering. We all go through life to experience and learn from both good and bad and positive and negative and positive and negative experiences, both joy and suffering and pain and joy, both good and bad and good and bad and negative and positive and negative and positive and negative and good and bad and the good and bad and negative and positive and negative and positive and negative and the good and bad and negative and positive and negative experiences, both good and bad and the good and bad and negative and the good and bad. The is the good and bad and the good and bad and the the good and bad and the the are the experiences. Through the joy and negative. The the the journey is the good. The through the the.
> ```

Great! The result is reasonable.

File: convert_checkpoints.py (hunk 5 above)
Comment (lsy323) [reply convert_checkpoints.py:407]: Yes, it's `freqs_cis`. Let's do not include the weight in safetensor, and reconstruct while loading

File: convert_checkpoints.py (hunk 4 above)
Comment (lsy323) [reply convert_checkpoints.py]: done

File: convert_checkpoints.py (hunk 1 above)
Comment (lsy323) [reply convert_checkpoints.py]: done

File: jetstream_pt/engine.py [hunk 6]
```diff
@@ -552,13 +552,38 @@ def _load_from_safetensors(self, path):
 
     return weights
 
+  def _load_from_state_dict(self, path):
+
+    state_dict = torch.load(path, map_location=torch.device("cpu"))[
+        "model_state_dict"
```
[the remark this answers is not in the document]
Comment (lsy323) [reply jetstream_pt/engine.py]: done

File: jetstream_pt/engine.py [hunk 7]
```diff
@@ -668,6 +693,9 @@ def create_pytorch_engine(
     raise NotImplementedError(
         "Loading from Pytorch raw checkpoint is not supported!"
     )
+  elif ".ckpt" in ckpt_path:
```
[the remark this answers is not in the document]
Comment (lsy323) [reply jetstream_pt/engine.py]: done

File: convert_checkpoints.py (hunk 2 above)
Comment (lsy323) [reply convert_checkpoints.py:420]: I opened a PR at https://huggingface.co/google/gemma-7b-pytorch/discussions/2

**Comment by lsy323** [comment]
> Can you also add the instructions in your PR to the README.md? thanks!

Added to the README.md for how to run Gemma.

**Comment by lsy323** [comment]
Updated `convert_checkpoint.py` to generate weight in `safetensor` format by default. Loading weights in `state_dict` is still added in the weight loading logic.

**Comment by FanhaiLu1** [comment]
> Updated `convert_checkpoint.py` to generate weight in `safetensor` format by default. Loading weights in `state_dict` is still added in the weight loading logic.

Great! All looks good to me now.

**Review by FanhaiLu1 — APPROVED** [review]

**Review by qihqi — APPROVED** [review]
